test flaky: TestControlPlane flaky running on OCP

Need to check flaky steps in the test:
```
  << Timeline
  [FAILED] Error getting sidecar version
  Unexpected error:
      <*fmt.wrapError | 0x4002643c60>: 
      error getting sidecar version: failed to get proxy status: error executing command: /usr/bin/sh -c /work/bin/istioctl proxy-status --namespace sample: Error: could not get XDS from discovery pod "istiod-default-v1-27-9-6cfb45ddfc-qnbkp": ADSC: failed running rpc error: code = Unavailable desc = connection error: desc = "transport: authentication handshake failed: EOF"
      
      {
          msg: "error getting sidecar version: failed to get proxy status: error executing command: /usr/bin/sh -c /work/bin/istioctl proxy-status --namespace sample: Error: could not get XDS from discovery pod \"istiod-default-v1-27-9-6cfb45ddfc-qnbkp\": ADSC: failed running rpc error: code = Unavailable desc = connection error: desc = \"transport: authentication handshake failed: EOF\"\n",
          err: <*fmt.wrapError | 0x4002643c20>{
              msg: "failed to get proxy status: error executing command: /usr/bin/sh -c /work/bin/istioctl proxy-status --namespace sample: Error: could not get XDS from discovery pod \"istiod-default-v1-27-9-6cfb45ddfc-qnbkp\": ADSC: failed running rpc error: code = Unavailable desc = connection error: desc = \"transport: authentication handshake failed: EOF\"\n",
              err: <*errors.errorString | 0x4001901760>{
                  s: "error executing command: /usr/bin/sh -c /work/bin/istioctl proxy-status --namespace sample: Error: could not get XDS from discovery pod \"istiod-default-v1-27-9-6cfb45ddfc-qnbkp\": ADSC: failed running rpc error: code = Unavailable desc = connection error: desc = \"transport: authentication handshake failed: EOF\"\n",
              },
          },
      }
  occurred
  In [It] at: /work/tests/e2e/controlplane/control_plane_update_test.go:207 @ 07/31/26 14:05:20.177
------------------------------
SSS•••••••••••••••••••••••••••••••••••••••••••••••
Summarizing 1 Failure:
  [FAIL] Control Plane updates using IstioRevisionTag v1.27.8 when the Istio CR is updated to the new Istio version [It] proxy version on sample pods still is base version [control-plane, slow]
  /work/tests/e2e/controlplane/control_plane_update_test.go:207
Ran 56 of 59 Specs in 387.158 seconds
FAIL! -- 55 Passed | 1 Failed | 0 Pending | 3 Skipped
--- FAIL: TestControlPlane (387.17s)
FAIL
```

--- a/tests/e2e/controlplane/control_plane_test.go
+++ b/tests/e2e/controlplane/control_plane_test.go
@@ -191,11 +191,13 @@ metadata:
 					})
 
 					It("has sidecars with the correct istio version", func(ctx SpecContext) {
-						for _, pod := range samplePods.Items {
-							sidecarVersion, err := common.GetProxyVersion(pod.Name, sampleNamespace)
-							Expect(err).NotTo(HaveOccurred(), "Error getting sidecar version")
-							Expect(sidecarVersion).To(Equal(version.Version), "Sidecar Istio version does not match the expected version")
-						}
+						Eventually(func(g Gomega) {
+							for _, pod := range samplePods.Items {
+								sidecarVersion, err := common.GetProxyVersion(pod.Name, sampleNamespace)
+								g.Expect(err).NotTo(HaveOccurred(), "Error getting sidecar version")
+								g.Expect(sidecarVersion).To(Equal(version.Version), "Sidecar Istio version does not match the expected version")
+							}
+						}).Should(Succeed(), "Error verifying sidecar version")
 						Success("Istio sidecar version matches the expected Istio version")
 					})
 				})

--- a/tests/e2e/controlplane/control_plane_update_test.go
+++ b/tests/e2e/controlplane/control_plane_update_test.go
@@ -188,9 +188,12 @@ spec:
 					Success("sample pods are ready")
 
 					for _, pod := range samplePods.Items {
-						sidecarVersion, err := common.GetProxyVersion(pod.Name, sampleNamespace)
-						Expect(err).NotTo(HaveOccurred(), "Error getting sidecar version")
-						Expect(sidecarVersion).To(Equal(baseVersion.Version), "Sidecar Istio version does not match the expected version")
+						podName := pod.Name
+						Eventually(func(g Gomega) {
+							sidecarVersion, err := common.GetProxyVersionFromPod(podName, sampleNamespace)
+							g.Expect(err).NotTo(HaveOccurred(), "Error getting sidecar version")
+							g.Expect(sidecarVersion).To(Equal(baseVersion.Version), "Sidecar Istio version does not match the expected version")
+						}).Should(Succeed(), "Error verifying sidecar version for pod "+podName)
 					}
 					Success("Istio sidecar version matches the expected base Istio version")
 				})
@@ -258,8 +261,12 @@ spec:
 					Expect(samplePods.Items).ToNot(BeEmpty(), "No pods found in sample namespace")
 
 					for _, pod := range samplePods.Items {
+						podName := pod.Name
 						Eventually(func(g Gomega) {
-							sidecarVersion, err := common.GetProxyVersion(pod.Name, sampleNamespace)
+							// Use GetProxyVersionFromPod to avoid XDS authentication issues
+							// that can occur during control plane upgrades when two istiod pods
+							// are running simultaneously on OCP.
+							sidecarVersion, err := common.GetProxyVersionFromPod(podName, sampleNamespace)
 							g.Expect(err).NotTo(HaveOccurred(), "Error getting sidecar version")
 							g.Expect(sidecarVersion).To(Equal(baseVersion.Version))
 						}).Should(Succeed(), "Sidecar Istio version does not match the expected version")

--- a/tests/e2e/migration/migration_procedure_test.go
+++ b/tests/e2e/migration/migration_procedure_test.go
@@ -449,19 +449,39 @@ spec:
 						Success("Pods running in ambient mode without sidecars")
 					})
 
-					It("verifies workloads are registered with ztunnel", func(ctx SpecContext) {
-						pods := &corev1.PodList{}
-						Expect(cl.List(ctx, pods, client.InNamespace(workloadNamespace))).To(Succeed())
-						Expect(pods.Items).NotTo(BeEmpty())
-
-						// Verify pods are detected by ztunnel
+					It("verifies CNI redirection is configured for workload pods", func(ctx SpecContext) {
+						// Checks that istio-cni has set the ambient.istio.io/redirection annotation,
+						// which means traffic interception rules for ztunnel are in place.
+						// End-to-end ztunnel registration is verified by the L4 connectivity test below.
 						Eventually(func(g Gomega) {
+							pods := &corev1.PodList{}
+							g.Expect(cl.List(ctx, pods, client.InNamespace(workloadNamespace))).To(Succeed())
+
+							workloadPods := []corev1.Pod{}
 							for _, pod := range pods.Items {
-								// Check pod annotations indicate ambient mode
-								g.Expect(pod.Annotations).To(HaveKey("ambient.istio.io/redirection"))
+								if pod.DeletionTimestamp != nil || pod.Status.Phase != corev1.PodRunning {
+									continue
+								}
+								isWaypoint := false
+								for key := range pod.Labels {
+									if key == "gateway.istio.io/managed" || key == "gateway.networking.k8s.io/gateway-name" {
+										isWaypoint = true
+										break
+									}
+								}
+								if !isWaypoint {
+									workloadPods = append(workloadPods, pod)
+								}
 							}
-						}).Should(Succeed())
-						Success("Workloads registered with ztunnel")
+
+							g.Expect(workloadPods).NotTo(BeEmpty(), "Expected at least one non-waypoint workload pod")
+							for _, pod := range workloadPods {
+								g.Expect(pod.Annotations).To(HaveKey("ambient.istio.io/redirection"),
+									"Pod %s should have ambient redirection annotation", pod.Name)
+							}
+						}).WithTimeout(3*time.Minute).WithPolling(5*time.Second).Should(Succeed(),
+							"All workload pods should have the ambient.istio.io/redirection annotation set by istio-cni")
+						Success("CNI redirection configured for all workload pods")
 					})
 
 					It("verifies L4 connectivity in ambient mode", func(ctx SpecContext) {

--- a/tests/e2e/operator/operator_install_test.go
+++ b/tests/e2e/operator/operator_install_test.go
@@ -43,6 +43,7 @@ import (
 	configv1 "github.com/openshift/api/config/v1"
 	crtls "github.com/openshift/controller-runtime-common/pkg/tls"
 	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
 	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/controller-runtime/pkg/client"
@@ -576,9 +577,47 @@ func ensureTLSAdherence(ctx context.Context, cl client.Client, policy configv1.T
 		return
 	}
 
+	// Wait for the operator to restart after the TLSAdherence change.
+	oldUIDs := make(map[string]struct{})
+	podList := &corev1.PodList{}
+	Expect(cl.List(ctx, podList, client.InNamespace(namespace),
+		client.MatchingLabels{"control-plane": deploymentName})).To(Succeed(),
+		"Failed to list operator pods before TLSAdherence change")
+	for _, pod := range podList.Items {
+		oldUIDs[string(pod.UID)] = struct{}{}
+	}
+
 	Step(fmt.Sprintf("Updating APIServer TLSAdherence to %s", policy))
 	apiServer.Spec.TLSAdherence = policy
 	Expect(cl.Update(ctx, apiServer)).To(Succeed(), "Failed to update APIServer TLSAdherence")
+
+	Step("Waiting for operator to restart and be ready after TLSAdherence change")
+	Eventually(func(g Gomega) {
+		pods := &corev1.PodList{}
+		g.Expect(cl.List(ctx, pods, client.InNamespace(namespace),
+			client.MatchingLabels{"control-plane": deploymentName})).To(Succeed())
+
+		found := false
+		for _, pod := range pods.Items {
+			if _, wasOld := oldUIDs[string(pod.UID)]; wasOld {
+				continue
+			}
+			if pod.DeletionTimestamp != nil || pod.Status.Phase != corev1.PodRunning {
+				continue
+			}
+			for _, cond := range pod.Status.Conditions {
+				if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+					found = true
+					break
+				}
+			}
+			if found {
+				break
+			}
+		}
+		g.Expect(found).To(BeTrue(), "No new ready operator pod found after TLSAdherence change to %s", policy)
+	}).WithTimeout(5*time.Minute).WithPolling(5*time.Second).Should(Succeed(),
+		"Operator should restart and be ready after TLSAdherence change")
 }
 
 func applyCustomTLSProfile(ctx context.Context, cl client.Client, ciphers []string) {

--- a/tests/e2e/util/common/e2e_utils.go
+++ b/tests/e2e/util/common/e2e_utils.go
@@ -472,6 +472,24 @@ func GetProxyVersion(podName, namespace string) (*semver.Version, error) {
 	return version, err
 }
 
+// GetProxyVersionFromPod extracts the Istio proxy version directly from the pod's
+// istio-proxy container by executing pilot-agent version. This approach does not
+// require XDS connectivity to istiod, making it more reliable during control plane
+// upgrades when istiod may be temporarily inaccessible.
+func GetProxyVersionFromPod(podName, namespace string) (*semver.Version, error) {
+	k := kubectl.New().WithNamespace(namespace)
+	output, err := k.Exec(podName, "istio-proxy", "pilot-agent version")
+	if err != nil {
+		return nil, fmt.Errorf("error getting proxy version from pod %s: %w", podName, err)
+	}
+
+	matches := istiodVersionRegex.FindStringSubmatch(output)
+	if len(matches) > 1 && matches[1] != "" {
+		return semver.NewVersion(matches[1])
+	}
+	return nil, fmt.Errorf("error getting proxy version from pod %s: version not found in output: %s", podName, output)
+}
+
 // GetIstioProxyContainer finds and returns the istio-proxy container from a pod
 // It checks both regular containers and init containers (for persistent init containers in K8s 1.28+)
 // Returns the container if found, nil otherwise

